Client/Frontend/Browser/Tabs/RemoteTabs/RemoteTabsPanel.swift

```diff
@@ -0,0 +1,379 @@
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        if refreshControl != nil {
+            removeRefreshControl()
+        }
+    }
+
+    func applyTheme() {
+        tableView.separatorColor = themeManager.currentTheme.colors.layerLightGrey30
+        if let delegate = tableViewDelegate as? RemoteTabsErrorDataSource {
+            delegate.applyTheme(theme: themeManager.currentTheme)
+        }
+    }
+
+    // MARK: - Refreshing TableView
+
+    func addRefreshControl() {
+        let control = UIRefreshControl()
+        control.addTarget(self, action: #selector(onRefreshPulled), for: .valueChanged)
+        refreshControl = control
+        tableView.refreshControl = control
+    }
+
+    func removeRefreshControl() {
+        tableView.refreshControl = nil
+        refreshControl = nil
+    }
+
+    @objc
+    func onRefreshPulled() {
+        refreshControl?.beginRefreshing()
+        refreshTabs(state: state, updateCache: true)
+    }
+
+    func endRefreshing() {
+        // Always end refreshing, even if we failed!
+        refreshControl?.endRefreshing()
+
+        // Remove the refresh control if the user has logged out in the meantime
+        if !state.allowsRefresh {
+            removeRefreshControl()
+        }
+    }
+
+    func updateDelegateClientAndTabData(_ clientAndTabs: [ClientAndTabs]) {
+        guard let remoteTabsPanel = remoteTabsPanel else { return }
+
+        guard !clientAndTabs.isEmpty else {
+            showEmptyTabsViewWith(.noClients)
+            return
+        }
+
+        let nonEmptyClientAndTabs = clientAndTabs.filter { !$0.tabs.isEmpty }
+        guard !nonEmptyClientAndTabs.isEmpty else {
+            showEmptyTabsViewWith(.noTabs)
+            return
+        }
+
+        let tabsPanelDataSource = RemoteTabsClientAndTabsDataSource(actionDelegate: remoteTabsPanel,
+                                                                    clientAndTabs: nonEmptyClientAndTabs,
+                                                                    theme: themeManager.currentTheme)
+        tabsPanelDataSource.collapsibleSectionDelegate = self
+        tableViewDelegate = tabsPanelDataSource
+
+        tableView.reloadData()
+    }
+
+    func refreshTabs(state: RemoteTabsPanelState, updateCache: Bool = false, completion: (() -> Void)? = nil) {
+        ensureMainThread { [self] in
+            self.state = state
+            performRefresh(updateCache: updateCache, completion: completion)
+        }
+    }
+
+    private func performRefresh(updateCache: Bool, completion: (() -> Void)?) {
+        // Short circuit if the user is not logged in
+        guard state.allowsRefresh else {
+            endRefreshing()
+            showEmptyTabsViewWith(.notLoggedIn)
+            return
+        }
+
+        // TODO: Send Redux action to get cached clients & tabs, update once new state is received. Forthcoming.
```

Comment:
I don't think we should merge this commented code into main yet mainly because is Redux work that should be part of future tickets like https://mozilla-hub.atlassian.net/browse/FXIOS-6943 and this calls in profile will live  in a middleware so it another separate ticket so a TODO will be enough here